### PR TITLE
Get latest 5 arrears actions

### PR DESCRIPTION
**Pull request by LBHRShetty**
Please review!

### CODE REVIEW

File: .idea/.idea.LBHTenancyAPI/.idea/indexLayout.xml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (StevenLeighton21) [inline .idea/.idea.LBHTenancyAPI/.idea/indexLayout.xml]: Should we add .idea to the `.gitignore`?

File: LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    public class ArrearsActionDiaryDetails
+    {
+        public string ActionCodeName { get; set; }
+        public string ActionCode { get; set; }
+        public DateTime ActionDate { get; set; }
+        public string ActionComment{ get; set; }
+        public decimal ActionBalance { get; set; }
+        public string UHUsername { get; set; }
+        public string TenancyRef{ get; set; }
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs]: Wild white space

File: LBHTenancyAPI/Domain/ArrearsAgreementDetail.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+        private string tenancyRef;
+        public string TenancyRef
+        {
+            get => tenancyRef;
+
+            set => tenancyRef = value.Trim();
+        }
+    }
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPI/Domain/ArrearsAgreementDetail.cs]: Wild white space, also I think we should clean this up a little bit (a few other whitespaces)

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs [hunk 4]
```diff
@@ -63,5 +64,68 @@ public List<TenancyListItem> GetTenanciesByRefs(List<string> tenancyRefs)
 
             return results;
         }
+
+        public Tenancy GetTenancyForRef(string tenancyRef)
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: Indentation is off - use `option + cmd + enter` in Rider

File: .idea/.idea.LBHTenancyAPI/.idea/indexLayout.xml (hunk 1 above)
Comment (StevenLeighton21) [reply .idea/.idea.LBHTenancyAPI/.idea/indexLayout.xml]: Also, remove .idea files from the commit - remove them with `git rm .idea/`

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs [hunk 5]
```diff
@@ -63,5 +64,68 @@ public List<TenancyListItem> GetTenanciesByRefs(List<string> tenancyRefs)
...
+                                       $"FROM tenagree " +
+                                       $"LEFT JOIN arag " +
+                                       $"ON arag.tag_ref = tenagree.tag_ref " +
+                                       $"LEFT JOIN contacts " +
+                                       $"ON contacts.tag_ref = tenagree.tag_ref " +
+                                       $"LEFT JOIN araction " +
+                                       $"ON araction.tag_ref = tenagree.tag_ref " +
+                                       $"WHERE tenagree.tag_ref = ('{tenancyRef}') " +
+                                       $"ORDER BY arag.arag_startdate DESC, araction.action_date DESC")
+                .FirstOrDefault();
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: I think we should keep consistent in how we write these queries, either keep the "$" or like the change above this change we should take the "$" out to keep the code consistent.

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs [hunk 6]
```diff
@@ -63,5 +64,68 @@ public List<TenancyListItem> GetTenanciesByRefs(List<string> tenancyRefs)
...
+                                                  $"action_code as ActionCode, " +
+                                                  $"action_code_name as ActionCodeName, " +
+                                                  $"action_date as ActionDate, " +
+                                                  $"action_comment as ActionComment, " +
+                                                  $"uh_username as UHUsername, " +
+                                                  $"action_balance as ActionBalance " +
+                                                  $"FROM araction " +
+                                                  $"WHERE tag_ref = ('{tenancyRef}') " +
+                                                  $"ORDER BY araction.action_date DESC").ToList();
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: Same here, we should take away the "$" so to keep the code consistent with the rest of the code.

File: LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+namespace LBHTenancyAPI.Domain
+{
+    public class ArrearsActionDiaryDetails
+    {
+        public string ActionCodeName { get; set; }
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs]: These should have custom setters that trim the value for `CHAR(x)` types, and the types of these fields must match the schema in UH.

You should also add a test showing that the data is as expected.

File: LBHTenancyAPI/LBHTenancyAPI.csproj.DotSettings [hunk 8]
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:String x:Key="/Default/CodeInspection/CSharpLanguageProject/LanguageLevel/@EntryValue">Experimental</s:String></wpf:ResourceDictionary>
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPI/LBHTenancyAPI.csproj.DotSettings]: Delete the space between + and <s:String .... >

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 9]
```diff
@@ -179,6 +179,54 @@ public void WhenGivenATenancyRefWithNoAddress_GetTenanciesByRefs_ShouldReturnNul
...
+            expectedTenancy.ArrearsAgreements = InsertRandomAgreementDetails(expectedTenancy.TenancyRef, 6);
+
+            var tenancy = GetSingleTenacyForRef(expectedTenancy.TenancyRef);
+
+            Assert.Equal(expectedTenancy.PrimaryContactName, tenancy.PrimaryContactName);
+            Assert.Equal(expectedTenancy.PrimaryContactPostcode, tenancy.PrimaryContactPostcode);
+            Assert.Equal(expectedTenancy.PrimaryContactLongAddress, tenancy.PrimaryContactLongAddress);
+            Assert.Equal(expectedTenancy.PrimaryContactPhone, tenancy.PrimaryContactPhone);
+        }
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: 3 wild white spaces

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 10]
```diff
@@ -214,12 +279,18 @@ private TenancyListItem InsertRandomisedTenancyListItem()
             return tenancy;
         }
 
-        private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Wild white space

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 11]
```diff
@@ -214,12 +279,18 @@ private TenancyListItem InsertRandomisedTenancyListItem()
...
+        private string InsertQueries()
         {
             string commandText =
                 "INSERT INTO tenagree (tag_ref, cur_bal) VALUES (@tenancyRef, @currentBalance);" +
-                "INSERT INTO contacts (tag_ref, con_name, con_address, con_postcode) VALUES (@tenancyRef, @primaryContactName, @primaryContactAddress, @primaryContactPostcode);";
+                "INSERT INTO contacts (tag_ref, con_name, con_address, con_postcode, con_phone1) VALUES (@tenancyRef, @primaryContactName, @primaryContactAddress, @primaryContactPostcode, @primaryContactPhone);";
 
+            return commandText;
+        }
+        private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Need to add an extra line here to separate the methods.

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 12]
```diff
@@ -243,6 +316,33 @@ private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
...
+                    : tenancyValues.PrimaryContactLongAddress + "\n";
+            command.Parameters.Add("@primaryContactPostcode", SqlDbType.Char);
+            command.Parameters["@primaryContactPostcode"].Value = tenancyValues.PrimaryContactPostcode;
+            command.Parameters.Add("@primaryContactPhone", SqlDbType.Char);
+            command.Parameters["@primaryContactPhone"].Value = tenancyValues.PrimaryContactPhone;
+
+
+            command.ExecuteNonQuery();
+
+            InsertArrearsActions(tenancyValues.TenancyRef, tenancyValues.LastActionCode,
```
Comment (StevenLeighton21) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: I would insert actions explicitly and separately here -

We need a test that shows the agreements and action lists are returned `null` when there are no agreements/actions.

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 13]
```diff
@@ -243,6 +316,33 @@ private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
...
+            command.Parameters["@primaryContactAddress"].Value =
+                tenancyValues.PrimaryContactLongAddress == null
+                    ? DBNull.Value.ToString()
+                    : tenancyValues.PrimaryContactLongAddress + "\n";
+            command.Parameters.Add("@primaryContactPostcode", SqlDbType.Char);
+            command.Parameters["@primaryContactPostcode"].Value = tenancyValues.PrimaryContactPostcode;
+            command.Parameters.Add("@primaryContactPhone", SqlDbType.Char);
+            command.Parameters["@primaryContactPhone"].Value = tenancyValues.PrimaryContactPhone;
+
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Wild white space

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 14]
```diff
@@ -259,6 +359,123 @@ private void InsertAgreement(string tenancyRef, string status, DateTime startDat
             command.ExecuteNonQuery();
         }
 
+
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: 2 wild white spaces

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 15]
```diff
@@ -259,6 +359,123 @@ private void InsertAgreement(string tenancyRef, string status, DateTime startDat
...
+                command.Parameters["@clearBy"].Value = arrearsAgreementDetail.ClearBy;
+
+                items.Add(arrearsAgreementDetail);
+                command.ExecuteNonQuery();
+            }
+
+            return items.OrderByDescending(i => i.Startdate).ToList();
+        }
+
+
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Wild white space

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 16]
```diff
@@ -259,6 +359,123 @@ private void InsertAgreement(string tenancyRef, string status, DateTime startDat
...
+                string commandText =
+
+                    "INSERT INTO araction (tag_ref, action_code,action_code_name,action_date,action_comment,action_balance," +
+                    "uh_username) " +
+                    "VALUES (@tenancyRef, @actionCode,@actionCodeName,@actionDate,@actionComment,@actionBalance,@uhUsername)";
+
+                foreach (int i in Enumerable.Range(0, num))
+                {
+                 ArrearsActionDiaryDetails arrearsActionDiaryDetail = new ArrearsActionDiaryDetails
+                 {
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: I don't think this is indented correctly

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 17]
```diff
@@ -259,6 +359,123 @@ private void InsertAgreement(string tenancyRef, string status, DateTime startDat
...
+
+                command.Parameters.Add("@uhUsername", SqlDbType.Char);
+                command.Parameters["@uhUsername"].Value = arrearsActionDiaryDetail.UHUsername;
+
+                items.Add(arrearsActionDiaryDetail);
+                command.ExecuteNonQuery();
+            }
+
+            return items.OrderByDescending(i => i.ActionDate).ToList();
+            }
```
Comment (VladyslavAtamanyuk) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: More indentation problems here

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 18]
```diff
@@ -259,6 +359,123 @@ private void InsertAgreement(string tenancyRef, string status, DateTime startDat
...
+            catch (Exception ex)
```
Comment (StevenLeighton21) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: I am not sure we need to catch an exception here - if an exception is thrown writing to the database, the tests will stop & fail anyway. This try/catch block just catches and re-throws the exception?

File: StubUniversalHousing/setup.sql [hunk 19]
```diff
@@ -3,7 +3,16 @@ GO
...
+                      action_date SMALLDATETIME,action_comment char(100),uh_username char(50));
+CREATE TABLE arag (tag_ref CHAR(11), arag_status CHAR(10), arag_startdate SMALLDATETIME,
+ arag_amount NUMERIC (9,3), arag_frequency CHAR(3), arag_breached BIT, arag_startbal NUMERIC (9,3), arag_clearby SMALLDATETIME);
 CREATE TABLE contacts (tag_ref CHAR(11), con_name VARCHAR(73), con_address CHAR(200), con_postcode CHAR(10), con_phone1 CHAR(21));
 GO
+
+select * from araction
+DROP TABLE arag
+
+select * from arag
```
Comment (StevenLeighton21) [inline StubUniversalHousing/setup.sql]: Rogue extra line, including dropping some tables that you need. These won't be run on setup but they should be removed.

File: StubUniversalHousing/setup.sql [hunk 20]
```diff
@@ -3,7 +3,16 @@ GO
...
+create TABLE araction (tag_ref CHAR(11), action_code CHAR(3),action_code_name char(50),action_balance decimal,
+                      action_date SMALLDATETIME,action_comment char(100),uh_username char(50));
+CREATE TABLE arag (tag_ref CHAR(11), arag_status CHAR(10), arag_startdate SMALLDATETIME,
+ arag_amount NUMERIC (9,3), arag_frequency CHAR(3), arag_breached BIT, arag_startbal NUMERIC (9,3), arag_clearby SMALLDATETIME);
 CREATE TABLE contacts (tag_ref CHAR(11), con_name VARCHAR(73), con_address CHAR(200), con_postcode CHAR(10), con_phone1 CHAR(21));
 GO
+
+select * from araction
+DROP TABLE arag
+
```
Comment (VladyslavAtamanyuk) [inline StubUniversalHousing/setup.sql]: Some cleaning needed here with some extra lines needed and also keep the SQL capitalised to keep the code readable and consistent.

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs [hunk 21]
```diff
@@ -243,6 +316,33 @@ private void InsertTenancyAttributes(TenancyListItem tenancyAttributes)
                 tenancyAttributes.LastActionDate);
         }
 
+        private void InsertSingleTenancyAttributes(Tenancy tenancyValues)
```
Comment (StevenLeighton21) [inline LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: Suggest renaming `tenancyValues` to `tenancyAttributes` for consistency with method name & other helpers.

File: LBHTenancyAPI/LBHTenancyAPI.csproj.DotSettings (hunk 8 above)
Comment (StevenLeighton21) [reply LBHTenancyAPI/LBHTenancyAPI.csproj.DotSettings]: I wouldn't modify generated files @VladyslavAtamanyuk - they'll only change back when they're regenerated.

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs (hunk 6 above)
Comment (StevenLeighton21) [reply LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: The `$` is needed only on the line with string interpolation, I am unsure of if we should retain the $ on all lines for consistency or use it only on lines we interpolate. I can see the argument either way.

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs (hunk 5 above)
Comment (StevenLeighton21) [reply LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: See below on similar comment

File: LBHTenancyAPI/LBHTenancyAPI.csproj.DotSettings (hunk 8 above)
Comment (VladyslavAtamanyuk) [reply LBHTenancyAPI/LBHTenancyAPI.csproj.DotSettings]: Ahh ok cool.

File: LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+        public string ActionCode { get; set; }
+        public DateTime ActionDate { get; set; }
+        public string ActionComment{ get; set; }
+        public decimal ActionBalance { get; set; }
+        public string UHUsername { get; set; }
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs]: Suggest `UhUsername`, or even better - `UniversalHousingUsername`

File: LBHTenancyAPI/Domain/Tenancy.cs [hunk 23]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace LBHTenancyAPI.Domain
+{
+    public struct Tenancy
+    {
+        public List<ArrearsAgreementDetail> ArrearsAgreements { get; set; }
+        public List<ArrearsActionDiaryDetails> ArrearsActionDiary { get; set; }
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Domain/Tenancy.cs]: Appreciating this is a change merged in that I wrote, there should probably be a newline separating these two lists, and no newline only for private fields used by custom accessors

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs [hunk 24]
```diff
@@ -63,5 +64,68 @@ public List<TenancyListItem> GetTenanciesByRefs(List<string> tenancyRefs)
...
+                                                      "FROM arag " +
+                                                      $"WHERE tag_ref = '{tencancyRef}'" +
+                                                      $"ORDER BY arag_startdate DESC ").ToList();
+        }
+
+            public List<ArrearsActionDiaryDetails> GetLatestfiveArrearsActionForRef(string tenancyRef)
+        {
+            return conn.Query<ArrearsActionDiaryDetails>($"" +
+                                                  $"SELECT top 5" +
+                                                  $"(tag_ref) as TenancyRef, " +
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: Don't need parentheses around `tag_ref`

File: LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs (hunk 7 above)
Comment (LBHRShetty) [reply LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs]: ok will do it.

File: LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs (hunk 22 above)
Comment (LBHRShetty) [reply LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs]: Done

File: LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs (hunk 2 above)
Comment (LBHRShetty) [reply LBHTenancyAPI/Domain/ArrearsActionDiaryDetails.cs]: removed wild whitespace

File: LBHTenancyAPI/Gateways/UhTenanciesGateway.cs (hunk 24 above)
Comment (LBHRShetty) [reply LBHTenancyAPI/Gateways/UhTenanciesGateway.cs]: removed them.

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 9 above)
Comment (LBHRShetty) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: done

File: LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs (hunk 17 above)
Comment (LBHRShetty) [reply LBHTenancyAPITest/Test/Gateways/UhTenanciesGatewayTest.cs]: done

File: StubUniversalHousing/setup.sql (hunk 19 above)
Comment (LBHRShetty) [reply StubUniversalHousing/setup.sql]: ok will do
